content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,433 @@
+from the
+[release package](https://github.com/llvm/llvm-project/releases) for your
+platform, or you can build it yourself.
+
+This is the minimal set of commands to build it yourself:
+```
+$ cmake -G Ninja ../llvm-project/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_PROJECTS="mlir"
+$ ninja tblgen-lsp-server
+```
+
+You will find `tblgen-lsp-server` in `<build-dir>/bin/`.
+
+The server reads a compilation database file, which is made for you when you
+configure LLVM using CMake.
+
+As long as your checkout of llvm-project includes
+[this commit](https://github.com/llvm/llvm-project/commit/c4afeccdd235a282d200c450e06a730504a66a08)
+the compilation database will include all TableGen files from all enabled
+projects (previously it was MLIR only).
+
+For example this configure will include information about TableGen files from
+llvm, clang, mlir and lldb:
+```
+$ cmake -G Ninja ../llvm-project/llvm -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_PROJECTS="clang;llvm;lldb;mlir"
+```
+
+This also applies to `-DLLVM_TARGETS_TO_BUILD=`. Enabling only one target will
+mean the compilation database only has files relevant to that target.
+
+**Note:** You do not need to build a project to include its TableGen files in
+the compilation database. Configuring is all that’s needed.
+
+Next, install the MLIR
+[extension](https://marketplace.visualstudio.com/items?itemName=llvm-vs-code-extensions.vscode-mlir)
+into Visual Studio Code (the language server can work with other but setup will
+differ).
+
+Then follow the setup instructions
+[here](https://mlir.llvm.org/docs/Tools/MLIRLSP/#td---tablegen-files) to tell
+the extension where the server and compilation database are.
+
+## Example
+
+Assuming you configured LLVM with the AArch64 target enabled (it is on by
+default).
+
+* Open the file `llvm/lib/Target/AArch64/AArch64.td`.
+* Put your cursor on a use of the `SubtargetFeature` type.
+* In the top menus select "Go" then "Go to Definition".
+* You will be taken to `llvm/include/llvm/Target/Target.td`, where
+  `SubtargetFeature` is defined.
+
+## Limitations
+
+The language server highlights an anti-pattern in the way some LLVM targets
+(e.g. AArch64) use TableGen.
+
+You may find yourself in a file that uses a class but does not define it, or
+include any files which do. This is because this file is intended to be included
+in another file, which does include a definition of that class.
+
+Perhaps we can address this by improving the language server, or reorganising
+the includes so we do not have these seemingly isolated files.
+
+# Dump
+
+What about printf? The best debugging tool of them all.
+
+TableGen’s equivalent is
+[dump](https://llvm.org/docs/TableGen/ProgRef.html#dump-print-messages-to-stderr),
+and its companion `repr`.
+```
+class Register<int _size, string _alias=""> {
+   int size = _size;
+   string alias = _alias;
+   dump "size is " # !repr(size) # " and alias is " # !repr(alias);
```

Comment:
I believe `string` and `int` can be represented already as strings when concatenated (last famous words, maybe you will prove me wrong). An example that does the dumping of a record might be more insightful?

Reply:
You're right, I've replaced the example with one adapted from the tests where repr makes a difference.

Also I found that this crashes:
```
class A <int _n, string _m> {
    int n = _n;
    string m = _m;
}

class B <A _a> {
    A a = _a;
    dump !repr(a);
    dump a;
}

def SomeB: B<A<1, "hello">>;
```
(https://godbolt.org/z/dKsd6qob5)

Reply:
I have created an issue for this. See https://github.com/llvm/llvm-project/issues/73094